providers/implementations/kdfs/argon2.c

```diff
@@ -0,0 +1,1488 @@
+/*
+ * Copyright 2017-2020 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdlib.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdarg.h>
+#include <limits.h>
+#include <string.h>
+#include <crypto/evp.h>
+#include <openssl/evp.h>
+#include <openssl/objects.h>
+#include <openssl/crypto.h>
+#include <openssl/kdf.h>
+#include <openssl/err.h>
+#include <openssl/core_names.h>
+#include <openssl/params.h>
+#include <internal/worker.h>
+#include <internal/numbers.h>
+#include <prov/implementations.h>
+#include <prov/provider_ctx.h>
+#include <prov/providercommonerr.h>
+#include <prov/blake2.h>
+
+#ifndef OPENSSL_NO_ARGON2
+
+# define ARGON2_MIN_LANES UINT32_C(1)
+# define ARGON2_MAX_LANES UINT32_C(0xFFFFFF)
+# define ARGON2_MIN_THREADS UINT32_C(1)
+# define ARGON2_MAX_THREADS UINT32_C(0xFFFFFF)
+# define ARGON2_SYNC_POINTS UINT32_C(4)
+# define ARGON2_MIN_OUTLEN UINT32_C(4)
+# define ARGON2_MAX_OUTLEN UINT32_C(0xFFFFFFFF)
+# define ARGON2_MIN_MEMORY (2 * ARGON2_SYNC_POINTS)
+# define ARGON2_MIN(a, b) ((a) < (b) ? (a) : (b))
+# define ARGON2_MAX_MEMORY_BITS                                                 \
+    ARGON2_MIN(UINT32_C(32), (sizeof(void *) * CHAR_BIT - 10 - 1))
+# define ARGON2_MAX_MEMORY                                                      \
+    ARGON2_MIN(UINT32_C(0xFFFFFFFF), UINT64_C(1) << ARGON2_MAX_MEMORY_BITS)
+# define ARGON2_MIN_TIME UINT32_C(1)
+# define ARGON2_MAX_TIME UINT32_C(0xFFFFFFFF)
+# define ARGON2_MIN_PWD_LENGTH UINT32_C(0)
+# define ARGON2_MAX_PWD_LENGTH UINT32_C(0xFFFFFFFF)
+# define ARGON2_MIN_AD_LENGTH UINT32_C(0)
+# define ARGON2_MAX_AD_LENGTH UINT32_C(0xFFFFFFFF)
+# define ARGON2_MIN_SALT_LENGTH UINT32_C(8)
```

Comment:
Maybe this should be 16?

Reply:
16 is recommended, not minimal (see https://github.com/P-H-C/phc-winner-argon2/blob/master/argon2-specs.pdf)